src/tools/miri/src/borrow_tracker/stacked_borrows/mod.rs

```diff
@@ -942,6 +942,11 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
 
                 // Check the type of this value to see what to do with it (retag, or recurse).
                 match place.layout.ty.kind() {
+                    &ty::Pat(base, pat) if *pat == ty::PatternKind::NotNull => {
+                        let mut place = place.clone();
+                        place.layout = self.ecx.layout_of(base)?;
+                        self.visit_value(&place)?;
+                    }
```

Comment:
had to special case it here to ensure we don't look at the components and try to retag a vtable pointer

Reply:
This is highly sus...

Reply:
No, something is wrong here. Pattern types should behave like ADTs with one field. The entire point of the shared value visitor machinery is that we don't have to repeat this kind of logic everywhere. So the retag visitor here should only have to spell out the types it cares about, and the fallback arm (which handles the recursive case) should work uniformly for everything else.

> it walks into the main pointer and into the vtable pointer and tries to retag both

What exactly is the type causing issues here?

Reply:
`*const dyn Trait is !null` is the type. The situation where it blows up is during `Retag` of such a value (the retag is generated around a `transmute` of a normal wide pointer to a pattern type).

I guess we can make it have one field, retaining the ScalarPair representation in the abi. I'll check that out